File: app/src/main/java/com/xuvjso/nfmovies/Utils/OkHttpUtil.java
package com.xuvjso.nfmovies.Utils;


import okhttp3.*;

import java.io.IOException;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.TimeUnit;

public class OkHttpUtil {
    public final static int READ_TIMEOUT = 10;
    public final static int CONNECT_TIMEOUT = 10;
    public final static int WRITE_TIMEOUT = 5;
    private static OkHttpUtil mInstance;
    private OkHttpClient mOkHttpClient;

    private OkHttpUtil() {
        okhttp3.OkHttpClient.Builder ClientBuilder = new okhttp3.OkHttpClient.Builder();
        ClientBuilder.readTimeout(READ_TIMEOUT, TimeUnit.SECONDS);
        ClientBuilder.writeTimeout(WRITE_TIMEOUT, TimeUnit.SECONDS);
        ClientBuilder.connectTimeout(CONNECT_TIMEOUT, TimeUnit.SECONDS);
        mOkHttpClient = ClientBuilder.build();
    }

    public static synchronized OkHttpUtil getInstance() {
        if (mInstance == null) {
            mInstance = new OkHttpUtil();
        }
        return mInstance;
    }

    public String getHtml(String url, String referer) {
        String html = null;
        Response response = getInstance().get(url, referer);
        if (response == null) return null;
        try {
            html = response.body().string();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return html;
    }

    public Response get(String url, String referer) {
        Request.Builder builder = new Request.Builder();
        Request request = builder.get().url(url)
                .addHeader("Referer", referer)
                .addHeader("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.70 Safari/537.36")
                .build();
        Call call = mOkHttpClient.newCall(request);
        Response response = null;
        try {
            response = call.execute();
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
        return response;
    }

    public Response post(String url, Map<String, String> params, String referer) {
        RequestBody body = setRequestBody(params);
        Request.Builder builder = new Request.Builder();
        Request request = builder.post(body).url(url).addHeader("Referer", referer).build();
        Call call = mOkHttpClient.newCall(request);
        Response response = null;
        try {
            response = call.execute();
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
        return response;
    }

    private RequestBody setRequestBody(Map<String, String> BodyParams) {
        RequestBody body = null;
        okhttp3.FormBody.Builder formEncodingBuilder = new okhttp3.FormBody.Builder();
        if (BodyParams != null) {
            Iterator<String> iterator = BodyParams.keySet().iterator();
            String key = "";
            while (iterator.hasNext()) {
                key = iterator.next().toString();
                formEncodingBuilder.add(key, BodyParams.get(key));
            }
        }
        body = formEncodingBuilder.build();
        return body;

    }

}

File: app/src/main/java/com/futuretech/nfmovies/Utils/OkHttpUtil.kt
package com.futuretech.nfmovies.Utils


import okhttp3.*

import java.io.IOException
import java.util.concurrent.TimeUnit

class OkHttpUtil private constructor() {
    private val mOkHttpClient: OkHttpClient

    init {
        val ClientBuilder = OkHttpClient.Builder()
        ClientBuilder.readTimeout(READ_TIMEOUT.toLong(), TimeUnit.SECONDS)
        ClientBuilder.writeTimeout(WRITE_TIMEOUT.toLong(), TimeUnit.SECONDS)
        ClientBuilder.connectTimeout(CONNECT_TIMEOUT.toLong(), TimeUnit.SECONDS)
        mOkHttpClient = ClientBuilder.build()
    }

    fun getHtml(url: String, referer: String): String? {
        var html: String? = null
        val response = instance[url, referer] ?: return null
        try {
            html = response.body!!.string()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return html
    }

    operator fun get(url: String, referer: String): Response? {
        val builder = Request.Builder()
        val request = builder.get().url(url)
                .addHeader("Referer", referer)
                .addHeader("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.70 Safari/537.36")
                .build()
        val call = mOkHttpClient.newCall(request)
        var response: Response? = null
        try {
            response = call.execute()
        } catch (e: IOException) {
            e.printStackTrace()
            return null
        }

        return response
    }

    fun post(url: String, params: Map<String, String>, referer: String): Response? {
        val body = setRequestBody(params)
        val builder = Request.Builder()
        val request = builder.post(body).url(url).addHeader("Referer", referer).build()
        val call = mOkHttpClient.newCall(request)
        var response: Response? = null
        try {
            response = call.execute()
        } catch (e: IOException) {
            e.printStackTrace()
            return null
        }

        return response
    }

    private fun setRequestBody(BodyParams: Map<String, String>?): RequestBody {
        val body: RequestBody?
        val formEncodingBuilder = FormBody.Builder()
        if (BodyParams != null) {
            val iterator = BodyParams.keys.iterator()
            var key : String
            while (iterator.hasNext()) {
                key = iterator.next()
                formEncodingBuilder.add(key, BodyParams.getValue(key))
            }
        }
        body = formEncodingBuilder.build()
        return body

    }

    companion object {
        const val READ_TIMEOUT = 10
        const val CONNECT_TIMEOUT = 10
        const val WRITE_TIMEOUT = 5
        private val mInstance: OkHttpUtil = OkHttpUtil()

        val instance: OkHttpUtil
            @Synchronized get() {
                return mInstance
            }
    }

}
